keps/sig-autoscaling/2702-graduate-hpa-api-to-GA/README.md

```diff
@@ -0,0 +1,426 @@
+Neither the deprecation of `v2beta1` nor the addition of `v2` requires any changes or conversion on the
+server side.  They will continue being stored in disk in v1 format as always, with new v2 fields serialized
+to annotiations.
+
+However any HPA objects stored in the user's code repository (all your YAML files) must stop using the
+v2beta1 format.  You should migrate all your HPA objects to the v2 format.  See the types.go files or just
+run `kubectl get hpa.v2.autoscaling -oyaml` to see your objects in the v2 format.
+
+###### Does enabling the feature change any default behavior?
+
+<!--
+Any change of default behavior may be surprising to users or break existing
+automations, so be extremely careful here.
+-->
+
+###### Can the feature be disabled once it has been enabled (i.e. can we roll back the enablement)?
+
+<!--
+Describe the consequences on existing workloads (e.g., if this is a runtime
+feature, can it break the existing applications?).
+
+NOTE: Also set `disable-supported` to `true` or `false` in `kep.yaml`.
+-->
+
+###### What happens if we reenable the feature if it was previously rolled back?
+
+###### Are there any tests for feature enablement/disablement?
+
+<!--
+The e2e framework does not currently support enabling or disabling feature
+gates. However, unit tests in each component dealing with managing data, created
+with and without the feature, are necessary. At the very least, think about
+conversion tests if API types are being modified.
+-->
+
+### Rollout, Upgrade and Rollback Planning
+
+<!--
+This section must be completed when targeting beta to a release.
+-->
+
+###### How can a rollout or rollback fail? Can it impact already running workloads?
+
+<!--
+Try to be as paranoid as possible - e.g., what if some components will restart
+mid-rollout?
+
+Be sure to consider highly-available clusters, where, for example,
+feature flags will be enabled on some API servers and not others during the
+rollout. Similarly, consider large clusters and how enablement/disablement
+will rollout across nodes.
+-->
+
+###### What specific metrics should inform a rollback?
+
+<!--
+What signals should users be paying attention to when the feature is young
+that might indicate a serious problem?
+-->
+
+###### Were upgrade and rollback tested? Was the upgrade->downgrade->upgrade path tested?
+
+<!--
+Describe manual testing that was done and the outcomes.
+Longer term, we may want to require automated upgrade/rollback tests, but we
+are missing a bunch of machinery and tooling and can't do that now.
+-->
+
+###### Is the rollout accompanied by any deprecations and/or removals of features, APIs, fields of API types, flags, etc.?
+
+<!--
+Even if applying deprecation policies, they may still surprise some users.
+-->
+
+### Monitoring Requirements
+
+<!--
+This section must be completed when targeting beta to a release.
+-->
+
+###### How can an operator determine if the feature is in use by workloads?
```

Comment:
> The question isn't whether it's supported, it's whether it's in use. So in this case the answer would be that some v2 resources exist. Not sure if we automatically have any metrics around presence or count of specific resource types.

@johnbelamaric All HPA objects are stored in v1 format on disk.  They are up converted the requested version.  And down converted upon update.

So there are no specifically versioned objects the user needs to worry about.  Am I understanding your question correctly?

Reply:
Ok, I suppose that makes sense. In that case, just the existence of HPA objects. We're sort of backfilling docs here for a long standing feature. In general as we see from the previous questions, I would prefer we answer the PRR from the point of view of "what will help cluster operators" with respect to this feature rather than from the point of view of "does specifically switching from v2beta2 to v2 have a specific impact with this question". That is, the questions are geared around the idea that going to GA means it's stable and folks will want to treat it as such, so "what does that mean for the operator?". How do they support it? How do they know it's working and if it's not working how do they fix the problem if it's not?

Given that this feature has been around a long time, you probably have all that knowledge, just not in one place. So taking it to GA gives us the opportunity to put it all in one place. Alternatively if you do have it in one place, a link there could suffice.

Reply:
We actually have quite good documentation IMO which has been updated as the newer v2 fields have been added: https://kubernetes.io/docs/tasks/run-application/horizontal-pod-autoscale/  This includes quite a bit of background, algorithm details, and some good [operator notes](https://kubernetes.io/docs/tasks/run-application/horizontal-pod-autoscale/#support-for-metrics-apis).

Reply:
Awesome. Please include links to that.